[Doc] Revamp ray core design patterns doc [18/n]: return ray put (#30318)
ray-project/ray: doc/source/ray-core/patterns/return-ray-put.rst

Anti-pattern: Returning ray.put() ObjectRefs from a task harms performance and fault tolerance
==============================================================================================

**TLDR:** Avoid calling :ref:`ray.put() <ray-put-ref>` on task return values and returning the resulting ObjectRefs;

> returning these values directly via :ref:`static <ray-task-returns>` and :ref:`dynamic <dynamic-generators>` ``num_returns``.

Reviewer: This is not a complete sentence. Also not clear why `num_returns` is relevant.

Returning ray.put() ObjectRefs are considered anti-patterns for the following reasons:

> - It disallows inlining small return values: Ray has a performance optimization to return small (<= 100KB) values inline directly to the owner, avoiding going through the distributed object store.

Reviewer: Avoid jargon like "owner".

Author: I can change to caller here. But later on we will discuss fault tolerance, should we also avoid "owner"?

  On the other hand, ``ray.put()`` will unconditionally store the value to the object store which makes the optimization for small return values impossible.
- Returning ObjectRefs involves extra distributed reference counting protocol which is slower than returning the values directly.
- It's less fault tolerant: the worker process that calls ``ray.put()`` is the owner of the return value and the return value fates share with the owner meaning that if the worker process dies, the return value is lost.
  In contrast, the caller process (often to be the driver) is the owner of the return value if it's returned directly.
  Fate sharing with the caller process is better than fate sharing with the callee process given that it's the caller that uses the return value.

.. note::

>   Currently actor tasks don't support dynamic returns so you have to use ``ray.put()`` to store each return value and return a list of ObjectRefs in this case.

Reviewer: This sentence doesn't really make sense because we haven't explained why you would want to use dynamic returns yet or what they are. We should explain the use case and how to replace `ray.put()` with generators first.

Anti-pattern: Returning ray.put() ObjectRefs from a task harms performance and fault tolerance
==============================================================================================

**TLDR:** Avoid calling :ref:`ray.put() <ray-put-ref>` on task return values and returning the resulting ObjectRefs.
Instead, return these values directly if possible.

Returning ray.put() ObjectRefs are considered anti-patterns for the following reasons:

- It disallows inlining small return values: Ray has a performance optimization to return small (<= 100KB) values inline directly to the caller, avoiding going through the distributed object store.
  On the other hand, ``ray.put()`` will unconditionally store the value to the object store which makes the optimization for small return values impossible.
- Returning ObjectRefs involves extra distributed reference counting protocol which is slower than returning the values directly.

> - It's less fault tolerant: the worker process that calls ``ray.put()`` is the owner of the return value and the return value fates share with the owner meaning that if the worker process dies, the return value is lost.
>   In contrast, the caller process (often to be the driver) is the owner of the return value if it's returned directly.
>   Fate sharing with the caller process is better than fate sharing with the callee process given that it's the caller that uses the return value.

Reviewer: ```suggestion
- It's less fault tolerant: the worker process that calls ``ray.put()`` is the "owner" of the returned ``ObjectRef`` and the return value fate shares with the owner. If the worker process dies, the return value is lost.
  In contrast, the caller process (often the driver) is the owner of the return value if it's returned directly.
```

Code example
------------

If you want to return a single value regardless if it's small or large, you should return it directly.

.. literalinclude:: ../doc_code/anti_pattern_return_ray_put.py
    :language: python
    :start-after: __return_single_value_start__
    :end-before: __return_single_value_end__

If you want to return multiple values and you know the number of returns before calling the task, you should use the :ref:`num_returns <ray-task-returns>` option.

.. literalinclude:: ../doc_code/anti_pattern_return_ray_put.py
    :language: python
    :start-after: __return_static_multi_values_start__
    :end-before: __return_static_multi_values_end__

> If you don't know the number of returns before calling the task, you should use the :ref:`dynamic generator <dynamic-generators>` if possible.

Reviewer: ```suggestion
If you don't know the number of returns before calling the task, you should use the :ref:`dynamic generator <dynamic-generators>` pattern if possible.
```
